torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -419,7 +424,10 @@ def _pre_backward_hook(
             _register_post_backward_final_callback(state)
             _clear_grads_if_needed(_all_handles(state))
         elif _handles_key:
-            _assert_in_training_states(state, [TrainingState.IDLE])
+            allowed_states = [TrainingState.IDLE]
+            if _is_composable(state):
```

Comment:
why are allowed states different in composable vs non-composable?

Reply:
First, note that this is `TrainingState`, which is per `state: _FSDPState` object, and not `HandleTrainingState`, which is per `FlatParamHandle` / `FlatParameter`. (`_FSDPState` is `Union[_State, FullyShardedDataParallel]`, where `_State` is from `torch/distributed/_composable/contract.py`).

For composable, `state` represents the local FSDP root (without wrapping). Upon the first `FlatParameter`'s pre-backward hook, the `state` will transition to `FORWARD_BACKWARD`. For any subsequent `FlatParameter`'s pre-backward hooks, `state.training_state` will already be in `FORWARD_BACKWARD`. However, each `FlatParamHandle`'s training state transitions like you expect (i.e. `IDLE` -> `BACKWARD_PRE` here).

This shows why I had to refactor `TrainingState` in an earlier PR. We have to stratify to accommodate the difference between `state: _FSDPState` and `FlatParamHandle` / `FlatParameter`.